pkg/client/fake/json_safe_client.go

```diff
@@ -0,0 +1,210 @@
+	}
+	return j.resource.Create(ctx, sanitized, options, subresources...)
+}
+
+func (j *jsonSafeResourceInterface) Update(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions, subresources ...string) (*unstructured.Unstructured, error) {
+	sanitized, err := sanitizeUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return j.resource.Update(ctx, sanitized, options, subresources...)
+}
+
+func (j *jsonSafeResourceInterface) UpdateStatus(ctx context.Context, obj *unstructured.Unstructured, options metav1.UpdateOptions) (*unstructured.Unstructured, error) {
+	sanitized, err := sanitizeUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return j.resource.UpdateStatus(ctx, sanitized, options)
+}
+
+func (j *jsonSafeResourceInterface) Delete(ctx context.Context, name string, options metav1.DeleteOptions, subresources ...string) error {
+	return j.resource.Delete(ctx, name, options, subresources...)
+}
+
+func (j *jsonSafeResourceInterface) DeleteCollection(ctx context.Context, options metav1.DeleteOptions, listOptions metav1.ListOptions) error {
+	return j.resource.DeleteCollection(ctx, options, listOptions)
+}
+
+func (j *jsonSafeResourceInterface) Get(ctx context.Context, name string, options metav1.GetOptions, subresources ...string) (*unstructured.Unstructured, error) {
+	return j.resource.Get(ctx, name, options, subresources...)
+}
+
+func (j *jsonSafeResourceInterface) List(ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error) {
+	return j.resource.List(ctx, opts)
+}
+
+func (j *jsonSafeResourceInterface) Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error) {
+	return j.resource.Watch(ctx, opts)
+}
+
+func (j *jsonSafeResourceInterface) Patch(ctx context.Context, name string, pt types.PatchType, data []byte, options metav1.PatchOptions, subresources ...string) (*unstructured.Unstructured, error) {
+	return j.resource.Patch(ctx, name, pt, data, options, subresources...)
+}
+
+func (j *jsonSafeResourceInterface) Apply(ctx context.Context, name string, obj *unstructured.Unstructured, options metav1.ApplyOptions, subresources ...string) (*unstructured.Unstructured, error) {
+	sanitized, err := sanitizeUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return j.resource.Apply(ctx, name, sanitized, options, subresources...)
+}
+
+func (j *jsonSafeResourceInterface) ApplyStatus(ctx context.Context, name string, obj *unstructured.Unstructured, options metav1.ApplyOptions) (*unstructured.Unstructured, error) {
+	sanitized, err := sanitizeUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return j.resource.ApplyStatus(ctx, name, sanitized, options)
+}
+
+// sanitizeUnstructured performs JSON marshal/unmarshal to sanitize unstructured data
+// This removes pointer references and makes the data safe for fake client deep copying
+func sanitizeUnstructured(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
```

Comment:
This is a deep copy, right?

maybe a better name for it?

Reply:
it is doing a deepcopy but it is attempting to remove `string*` from the unstructured so it does a bit more. This only matters for the fake dynamic client, it does not understand how to deep copy a `string*` in the conditions

Reply:
Right that makes sense. My question was specifically for the fake impl - you'd think they would deepCopy in the fake impl to save us having to do this was my point.

Reply:
renamed to `deepCopyUnstructured`

Reply:
I don't think they do because the unstructured object does not have the member structs that have deepcopy attached and they do a reflection based copy field by field.